llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -6297,6 +6297,14 @@ SIInstrInfo::legalizeOperands(MachineInstr &MI,
     return CreatedBB;
   }
 
+  // Legalize S_BITREPLICATE
+  if (MI.getOpcode() == AMDGPU::S_BITREPLICATE_B64_B32) {
+    MachineOperand &Src = MI.getOperand(1);
+    if (Src.isReg() && RI.hasVectorRegisters(MRI.getRegClass(Src.getReg())))
+      Src.setReg(readlaneVGPRToSGPR(Src.getReg(), MI, MRI));
+    return CreatedBB;
+  }
```

Comment:
This unfortunately seems necessary.

The FixSGPRCopies pass incorrectly decides to turn `s_bitreplicate` into a VALU instruction when there is a VGPR input. This is a similar issue to [D45826](https://reviews.llvm.org/D45826).

I refrained from writing a helper method for now because there would be only two calls for now. If it makes sense to add one, I will.

Reply:
Emitting readfirstlane is just wrong, you have to emulate the operation with a VALU expansion if needed

Reply:
> Emitting readfirstlane is just wrong, you have to emulate the operation with a VALU expansion if needed

Even with the `convergent` attribute of the intrinsic? I thought we can rule out any divergent inputs with that.

Reply:
No, that just maybe will prevent the optimizer from introducing new divergences. You cannot guarantee any arbitrary input was uniform from the start

Reply:
There’s more discussions and reasoning about this in the main thread of this PR, but as far as I understand the intention, the intrinsic only accepts uniform values. From the intrinsic definition:
```
The argument must be uniform; otherwise, the result is undefined.
```

A uniform value can still end up in a VGPR, so it makes sense to me to use readfirstlane in that case.